crates/miden-objects/src/address/mod.rs

```diff
@@ -1,364 +1,230 @@
 mod r#type;
+use alloc::string::ToString;
+
+pub use r#type::AddressType;
+
+mod routing_parameters;
+use alloc::borrow::ToOwned;
+
+pub use routing_parameters::RoutingParameters;
 
 mod interface;
 mod network_id;
-use alloc::string::{String, ToString};
+use alloc::string::String;
 
-use bech32::Bech32m;
-use bech32::primitives::decode::{ByteIter, CheckedHrpstring};
 pub use interface::AddressInterface;
-use miden_core::utils::{ByteWriter, Deserializable, Serializable};
 use miden_processor::DeserializationError;
 pub use network_id::{CustomNetworkId, NetworkId};
-pub use r#type::AddressType;
 
 use crate::AddressError;
-use crate::account::{AccountId, AccountStorageMode};
-use crate::errors::Bech32Error;
+use crate::account::AccountStorageMode;
 use crate::note::NoteTag;
+use crate::utils::serde::{ByteWriter, Deserializable, Serializable};
+
+mod address_id;
+pub use address_id::AddressId;
 
 /// A user-facing address in Miden.
-#[non_exhaustive]
+///
+/// An address consists of an [`AddressId`] and optional [`RoutingParameters`].
+///
+/// A user who wants to receive a note creates an address and sends it to the sender of the note.
+/// The sender creates a note that only the address ID can consume (i.e. it provides access-control)
+/// and the routing parameters inform the sender about various aspects like:
+/// - what kind of note the receiver's account can consume.
+/// - how the receiver discovers the note.
+///
+/// It can be encoded to a string using [`Self::encode`] and decoded using [`Self::decode`].
+/// If routing parameters are present, the ID and parameters are separated by
+/// [`Address::SEPARATOR`].
+///
+/// ## Example
+///
+/// ```text
+/// # account ID
+/// mm1qrgu2vdxjqt9curephr0c94n0quevadt
+/// # account ID + routing parameters
+/// mm1qrgu2vdxjqt9curephr0c94n0quevadt_qrcqzlvsfdp
```

Comment:
Question: can we somehow impose structure for the character that comes after the separator? Specifically, would it be difficult to set things up so that if the ID is account ID, the address starts as `mm1a`.

Then for public key based addresses, we could do something like `mm1k` (or maybe `mm1p`).

This is a very minor thing - so, if it introduces non-negligible complexity, let's leave things as they are now.

Reply:
Yes, great suggestion! I changed the `AddressType::AccountId` to `232` which means the first 5 bits encode character `a`.